modules/internal/ChapelDomain.chpl

```diff
@@ -1283,6 +1283,271 @@ module ChapelDomain {
       return _newArray(x);
     }
 
+    /*
+      This manager is returned by ``unsafeAssign()`` and can be used in
+      managed blocks to help resize arrays of non-default-initializable
+      elements.
+    */
+    record unsafeAssignManager {
+      pragma "no doc"
+      var _lhsInstance;
+
+      pragma "no doc"
+      var _lhsPid: int;
+
+      pragma "no doc"
+      var _rhsInstance;
+
+      pragma "no doc"
+      var _rhsPid: int;
+
+      pragma "no doc"
+      var _oldLhsDomainCopy: domain;
+
+      pragma "no doc"
+      param _checks: bool;
+
+      pragma "no doc"
+      var _isActiveManager: bool;
+
+      pragma "no doc"
+      iter _lhsBaseArrays() {
+        for arr in _lhsInstance._arrs do yield arr;
+      }
+
+      /*
+        Returns 'true' if this manager has runtime safety checks enabled.
+      */
+      inline proc checks return _checks;
+
+      /*
+        Check if a given class reference is ``nil`` without triggering
+        runtime nilability checks.
+      */
+      proc isClassReferenceNil(const ref x) {
+        if isClassType(x.type) {
+          return if isBorrowedOrUnmanagedClassType(x.type)
+            then x == nil
+            else x.chpl_p == nil;
+        } else {
+          return false;
+        }
+      }
+
+      /*
+        Return ``true`` if the value at a given index in an array has
+        been initialized.
+      */
+      proc isElementInitialized(arr: [?d], idx) {
+        if _checks {
+          type T = arr.eltType;
+          if isNonNilableClassType(T) {
+            var actualArr = chpl__getActualArray(arr);
+            return !actualArr.chpl_unsafeAssignIsClassElementNil(this, idx);
+          } else if !isDefaultInitializable(T) {
+            halt('Not implemented yet!');
```

Comment:
In case someone bumps into this in the interim, could you make the message more clear?